Adds plugin installation functionality

johnSchnake opened the pull request:
Adds a known location, ~/.sonobuoy or SONOBUOY_DIR env,
which stores plugin files (*.yaml). When running plugins this location is searched
first so that users can more easily call custom plugins from any location.

Adds:
 - plugins list command
 - plugin show command
 - plugin install command
 - plugin uninstall command
 - tests for the above

Partially fixes: #1292
Signed-off-by: John Schnake <jschnake@vmware.com>

johnSchnake commented:
Have basic list/show/install commands with an env override for the installation location.

TODO:
 - [ ] lots of tests
 - [ ] clarify help messages for usage
 - [ ] rework slightly so the methods dont output the CLI output. Ensure the functional code is in its own methods and the CLI stuff is separate.
 - [ ] more? Recheck the DD

johnSchnake commented:
Also added an uninstall command and am making the commands hidden for now. This way all this can merge before actually tying it into the actual plugin loading. So this is just a stealth feature a.t.m. where sonobuoy will save/list/show plugins but it doesn't impact runs at all.

The next PR can actually tie it into the loading of plugins (thats the part that we'll feature gate with an env var).

johnSchnake commented:
@vladimirvivien Since these are hidden commands at first we don't have to be perfect but I wanted eyes on this first main chunk of the functionality. Anything that doesn't make sense or that I failed to implement from the DD?

The followup PR will make these non-hidden and actually modify the loading of plugins. The point of merging this first is to get the bulk of the code out of the way so we can more easily code/review the implementation and tweaks on this foundational PR.

vladimirvivien commented on cmd/sonobuoy/app/pluginCache.go:
```diff
@@ -0,0 +1,266 @@
+	if len(usePath) == 0 {
+		usePath = defaultSonobuoyDir
+	}
+	expandedPath, err := homedir.Expand(usePath)
```
Suggestion: you could save a dependency here. If I recall correctly, that `home-dir` package essentially does the following.

```go
func ExpandPath(path string) (string, error) {
	home, err := os.UserHomeDir()
	if err != nil {
		return "", err
	}

	if path[0] == '~' {
		path = filepath.Join(home, path[1:])
	}
	return path, nil
}
```

johnSchnake replied:
Maybe I will; originally I used home-dir since it mentioned something about cgo but that was very outdated. https://github.com/mitchellh/go-homedir/issues/20#issuecomment-467476994 mentions the stdlib vs home-dir usage and I may just prefer the stdlib in our case.

vladimirvivien commented on cmd/sonobuoy/app/pluginCache.go:
```diff
@@ -0,0 +1,266 @@
+	if _, err := os.Stat(expandedPath); err != nil && os.IsNotExist(err) {
+		logrus.Debugf("sonobuoy plugin location %q does not exist, creating it.", expandedPath)
+		if err := os.Mkdir(expandedPath, 0777); err != nil {
+			logrus.Errorf("failed to create directory for installed plugins %q: %v", expandedPath, err)
```
Should this if-block `return ""` dir failed to be created ?

vladimirvivien commented on cmd/sonobuoy/app/pluginCache.go:
```diff
@@ -0,0 +1,266 @@
+// loadPlugins loads all plugins from the given directory (recursively) and
+// returns a map of their filename and contents. Returns the first error encountered
+// but continues to load and return as many plugins as possible.
+func loadPlugins(installedDir string) (map[string]*manifest.Manifest, error) {
```
Suggestion: maybe this should return a `[]error` to collect all errs as they happen (since the walk continues). Not sure if that would be useful for the caller of this func .

johnSchnake replied:
I'm always a little split on that but I tend to usually prefer the 'returns first error' since programatically, in most cases, we just need to know error/pass and the user still gets logs for each error. Makes it easier/prettier to handle rather than a []error to handle.

vladimirvivien commented on cmd/sonobuoy/app/pluginCache.go:
```diff
@@ -0,0 +1,266 @@
+	return pluginMap, firstErr
+}
+
+func loadPlugin(installedDir, reqFile string) (*manifest.Manifest, error) {
```
Could this be done from the pre-loaded `map[string]*manifest.Manifest` from above ?

johnSchnake replied:
I thought about doing it that way too; it could definitely leverage it if we wanted but the main difference was simply that this exits early if found. That's not a huge issue since users won't have 1 million plugins or anything. You're probably right that the simplification is preferred.

vladimirvivien reviewed (commented):
This looks great @johnSchnake 
Left a few comments/suggestions.

johnSchnake commented:
Fixed up from comments and added a new test to confirm that when listing plugins we continue despite failures (which we didnt actually do before).

Merging and working on the followup that actually ties this into sonobuoy logic
